planner/core/optimizer.go

```diff
@@ -343,11 +364,43 @@ func logicalOptimize(ctx context.Context, flag uint64, logic LogicalPlan) (Logic
 		if flag&(1<<uint(i)) == 0 || isLogicalRuleDisabled(rule) {
 			continue
 		}
-		logic, err = rule.optimize(ctx, logic)
+		logicalOptimizeReturn, err := rule.optimize(ctx, logic)
+		if err != nil {
+			return nil, err
+		}
+		logic = logicalOptimizeReturn.logicalPlan
+	}
+	return logic, err
+}
+
+func logicalOptimizeWithTraceEnabled(ctx context.Context, flag uint64, logic LogicalPlan) (LogicalPlan, error) {
+	trace := &tracing.LogicalOptimizeTrace{
+		Steps: make(map[int]tracing.LogicalRuleOptimizeTrace),
+	}
+	var err error
+	for i, rule := range optRuleList {
+		// The order of flags is same as the order of optRule in the list.
+		// We use a bitmask to record which opt rules should be used. If the i-th bit is 1, it means we should
+		// apply i-th optimizing rule.
+		if flag&(1<<uint(i)) == 0 || isLogicalRuleDisabled(rule) {
```

Comment:
This part is duplicated with the original one, you can align. Please think about minimizing changes to original codes to make tracer work.

Reply:
updated.